server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,136 @@
+package pr
+
+import (
+	internalContext "github.com/runatlantis/atlantis/server/neptune/context"
+	workflowMetrics "github.com/runatlantis/atlantis/server/neptune/workflows/internal/metrics"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/pr/revision"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/plugins"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+type Action int64
+
+const (
+	onNewRevision Action = iota
+	onShutdown
+)
+
+type RunnerState int64
+
+const (
+	working RunnerState = iota
+	waiting
+	complete
+)
+
+const ShutdownSignalID = "pr-close"
+
+type NewShutdownRequest struct{}
+
+type RevisionProcessor interface {
+	Process(ctx workflow.Context, prRevision revision.Revision)
+}
+
+type Runner struct {
+	RevisionSignalChannel workflow.ReceiveChannel
+	RevisionReceiver      *revision.Receiver
+	ShutdownSignalChannel workflow.ReceiveChannel
+	RevisionProcessor     RevisionProcessor
+	Scope                 workflowMetrics.Scope
+
+	// mutable state
+	state                 RunnerState
+	lastAttemptedRevision string
+	cancel                workflow.CancelFunc
+}
+
+func newRunner(ctx workflow.Context, scope workflowMetrics.Scope, tfWorkflow revision.TFWorkflow, internalNotifiers []revision.WorkflowNotifier, additionalNotifiers ...plugins.TerraformWorkflowNotifier) *Runner {
+	revisionReceiver := revision.NewRevisionReceiver(ctx, scope)
+	stateReceiver := revision.StateReceiver{
+		InternalNotifiers:   internalNotifiers,
+		AdditionalNotifiers: additionalNotifiers,
+	}
+	revisionProcessor := revision.Processor{
+		TFWorkflow:      tfWorkflow,
+		TFStateReceiver: &stateReceiver,
+		PolicyHandler:   &revision.FailedPolicyHandler{},
+	}
+	return &Runner{
+		RevisionSignalChannel: workflow.GetSignalChannel(ctx, revision.TerraformRevisionSignalID),
+		RevisionReceiver:      &revisionReceiver,
+		ShutdownSignalChannel: workflow.GetSignalChannel(ctx, ShutdownSignalID),
+		Scope:                 scope,
+		RevisionProcessor:     &revisionProcessor,
+
+		cancel: func() {},
+	}
+}
+
+// Run handles managing the workflow's context lifecycles as new signals/poll events are received and
+// change the current PRAction status
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action Action
+	var prRevision revision.Revision
+
+	//TODO: add approve signal, timeout, poll variation of shutdown signal
+	s := workflow.NewSelector(ctx)
+	s.AddReceive(r.RevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		prRevision = r.RevisionReceiver.Receive(c, more)
+		action = onNewRevision
+	})
+	s.AddReceive(r.ShutdownSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		defer func() {
+			action = onShutdown
+			r.state = complete
+		}()
+		if !more {
+			return
+		}
+		ctx = workflow.WithRetryPolicy(ctx, temporal.RetryPolicy{
+			MaximumAttempts: 5,
+		})
```

Comment:
Whats the point of this?

Reply:
copypasta, it's what deploy workflow uses when receiving a signal for a new revision. I can remove from here, especially since we also want to poll for shutdown events